// @nextui-org/snippet/dist/chunk-Y3JJ2477.mjs
"use client";

// src/use-snippet.ts
import { snippet } from "@nextui-org/theme";
import { mapPropsVariants } from "@nextui-org/system";
import { useDOMRef } from "@nextui-org/react-utils";
import { clsx, dataAttr } from "@nextui-org/shared-utils";
import { useClipboard } from "@nextui-org/use-clipboard";
import { useFocusRing } from "@react-aria/focus";
import { useMemo, useCallback, useRef } from "react";
function useSnippet(originalProps) {
  const [props, variantProps] = mapPropsVariants(originalProps, snippet.variantKeys);
  const {
    ref,
    as,
    children,
    symbol = "$",
    classNames,
    timeout,
    copyIcon,
    checkIcon,
    codeString,
    disableCopy = false,
    disableTooltip = false,
    hideCopyButton = false,
    autoFocus = false,
    hideSymbol = false,
    onCopy: onCopyProp,
    tooltipProps = {
      offset: 15,
      delay: 1e3,
      content: "Copy to clipboard",
      color: originalProps == null ? void 0 : originalProps.color,
      isDisabled: disableCopy
    },
    copyButtonProps,
    className,
    ...otherProps
  } = props;
  const Component = as || "div";
  const domRef = useDOMRef(ref);
  const preRef = useRef(null);
  const { copy, copied } = useClipboard({ timeout });
  const isMultiLine = children && Array.isArray(children);
  const { isFocusVisible, isFocused, focusProps } = useFocusRing({
    autoFocus
  });
  const slots = useMemo(
    () => snippet({
      ...variantProps
    }),
    [...Object.values(variantProps)]
  );
  const symbolBefore = useMemo(() => {
    if (!symbol || typeof symbol !== "string")
      return symbol;
    const str = symbol.trim();
    return str ? `${str} ` : "";
  }, [symbol]);
  const baseStyles = clsx(classNames == null ? void 0 : classNames.base, className);
  const getSnippetProps = useCallback(
    () => ({
      className: slots.base({
        class: baseStyles
      }),
      ...otherProps
    }),
    [slots, baseStyles, isMultiLine, otherProps]
  );
  const onCopy = useCallback(() => {
    var _a;
    if (disableCopy) {
      return;
    }
    let stringValue = "";
    if (typeof children === "string") {
      stringValue = children;
    } else if (Array.isArray(children)) {
      children.forEach((child) => {
        var _a2, _b;
        const childString = typeof child === "string" ? child : (_b = (_a2 = child == null ? void 0 : child.props) == null ? void 0 : _a2.children) == null ? void 0 : _b.toString();
        if (childString) {
          stringValue += childString + "\n";
        }
      });
    }
    const valueToCopy = codeString || stringValue || ((_a = preRef.current) == null ? void 0 : _a.textContent) || "";
    copy(valueToCopy);
    onCopyProp == null ? void 0 : onCopyProp(valueToCopy);
  }, [copy, codeString, disableCopy, onCopyProp, children]);
  const defaultCopyButtonProps = {
    "aria-label": typeof tooltipProps.content === "string" ? tooltipProps.content : "Copy to clipboard",
    size: "sm",
    variant: "light",
    isDisabled: disableCopy,
    onPress: onCopy,
    isIconOnly: true,
    ...copyButtonProps
  };
  const getCopyButtonProps = useCallback(
    () => ({
      ...defaultCopyButtonProps,
      "data-copied": dataAttr(copied),
      className: slots.copyButton({
        class: clsx(classNames == null ? void 0 : classNames.copyButton)
      })
    }),
    [
      slots,
      isFocusVisible,
      isFocused,
      disableCopy,
      classNames == null ? void 0 : classNames.copyButton,
      defaultCopyButtonProps,
      focusProps
    ]
  );
  return {
    Component,
    as,
    domRef,
    preRef,
    children,
    slots,
    classNames,
    copied,
    onCopy,
    copyIcon,
    checkIcon,
    symbolBefore,
    isMultiLine,
    isFocusVisible,
    hideCopyButton,
    disableCopy,
    disableTooltip,
    hideSymbol,
    tooltipProps,
    getSnippetProps,
    getCopyButtonProps
  };
}

export {
  useSnippet
};
