// @nextui-org/theme/dist/chunk-PW52XQD7.mjs
import {
  createSpacingUnits,
  generateSpacingScale,
  isBaseTheme
} from "./chunk-NW5W5ANO.mjs";
import {
  utilities
} from "./chunk-DYIIRM5B.mjs";
import {
  baseStyles
} from "./chunk-T7JNS25F.mjs";
import {
  semanticColors
} from "./chunk-G4RCK475.mjs";
import {
  flattenThemeObject
} from "./chunk-KUNVFLXJ.mjs";
import {
  darkLayout,
  defaultLayout,
  lightLayout
} from "./chunk-7MCHEQXT.mjs";
import {
  animations
} from "./chunk-DMASP6FA.mjs";
import {
  commonColors
} from "./chunk-IAS3SFA4.mjs";

// src/plugin.ts
import Color from "color";
import plugin from "tailwindcss/plugin.js";
import get from "lodash.get";
import omit from "lodash.omit";
import forEach from "lodash.foreach";
import mapKeys from "lodash.mapkeys";
import kebabCase from "lodash.kebabcase";
import deepMerge from "deepmerge";
var DEFAULT_PREFIX = "nextui";
var resolveConfig = (themes = {}, defaultTheme, prefix) => {
  const resolved = {
    variants: [],
    utilities: {},
    colors: {}
  };
  forEach(themes, ({ extend, layout, colors }, themeName) => {
    let cssSelector = `.${themeName},[data-theme="${themeName}"]`;
    const scheme = themeName === "light" || themeName === "dark" ? themeName : extend;
    if (themeName === defaultTheme) {
      cssSelector = `:root,${cssSelector}`;
    }
    resolved.utilities[cssSelector] = scheme ? {
      "color-scheme": scheme
    } : {};
    const flatColors = flattenThemeObject(colors);
    const flatLayout = layout ? mapKeys(layout, (value, key) => kebabCase(key)) : {};
    resolved.variants.push({
      name: themeName,
      definition: [`&.${themeName}`, `&[data-theme='${themeName}']`]
    });
    forEach(flatColors, (colorValue, colorName) => {
      if (!colorValue)
        return;
      try {
        const [h, s, l, defaultAlphaValue] = Color(colorValue).hsl().round().array();
        const nextuiColorVariable = `--${prefix}-${colorName}`;
        const nextuiOpacityVariable = `--${prefix}-${colorName}-opacity`;
        resolved.utilities[cssSelector][nextuiColorVariable] = `${h} ${s}% ${l}%`;
        if (typeof defaultAlphaValue === "number") {
          resolved.utilities[cssSelector][nextuiOpacityVariable] = defaultAlphaValue.toFixed(2);
        }
        resolved.colors[colorName] = ({ opacityVariable, opacityValue }) => {
          if (!isNaN(+opacityValue)) {
            return `hsl(var(${nextuiColorVariable}) / ${opacityValue})`;
          }
          if (opacityVariable) {
            return `hsl(var(${nextuiColorVariable}) / var(${nextuiOpacityVariable}, var(${opacityVariable})))`;
          }
          return `hsl(var(${nextuiColorVariable}) / var(${nextuiOpacityVariable}, 1))`;
        };
      } catch (error) {
        console.log("error", error == null ? void 0 : error.message);
      }
    });
    forEach(flatLayout, (value, key) => {
      if (!value)
        return;
      if (typeof value === "object") {
        forEach(value, (v, k) => {
          const layoutVariable = `--${prefix}-${key}-${k}`;
          resolved.utilities[cssSelector][layoutVariable] = v;
        });
      } else if (key === "spacing-unit") {
        const layoutVariable = `--${prefix}-${key}`;
        resolved.utilities[cssSelector][layoutVariable] = value;
        const spacingScale = generateSpacingScale(Number(value));
        forEach(spacingScale, (v, k) => {
          const layoutVariable2 = `--${prefix}-${key}-${k}`;
          resolved.utilities[cssSelector][layoutVariable2] = v;
        });
      } else {
        const layoutVariable = `--${prefix}-${key}`;
        resolved.utilities[cssSelector][layoutVariable] = value;
      }
    });
  });
  return resolved;
};
var corePlugin = (themes = {}, defaultTheme, prefix, addCommonColors) => {
  const resolved = resolveConfig(themes, defaultTheme, prefix);
  const minSizes = {
    "unit-1": `var(--${prefix}-spacing-unit)`,
    "unit-2": `var(--${prefix}-spacing-unit-2`,
    "unit-3": `var(--${prefix}-spacing-unit-3)`,
    "unit-3.5": `var(--${prefix}-spacing-unit-3_5)`,
    "unit-4": `var(--${prefix}-spacing-unit-4)`,
    "unit-5": `var(--${prefix}-spacing-unit-5)`,
    "unit-6": `var(--${prefix}-spacing-unit-6)`,
    "unit-7": `var(--${prefix}-spacing-unit-7)`,
    "unit-8": `var(--${prefix}-spacing-unit-8)`,
    "unit-9": `var(--${prefix}-spacing-unit-9)`,
    "unit-10": `var(--${prefix}-spacing-unit-10)`,
    "unit-11": `var(--${prefix}-spacing-unit-11)`,
    "unit-12": `var(--${prefix}-spacing-unit-12)`,
    "unit-16": `var(--${prefix}-spacing-unit-16)`,
    "unit-20": `var(--${prefix}-spacing-unit-20)`,
    "unit-24": `var(--${prefix}-spacing-unit-24)`
  };
  return plugin(
    ({ addBase, addUtilities, addVariant }) => {
      addBase({
        [":root, [data-theme]"]: {
          ...baseStyles(prefix)
        }
      });
      addUtilities({ ...resolved.utilities, ...utilities });
      resolved.variants.forEach((variant) => {
        addVariant(variant.name, variant.definition);
      });
    },
    {
      theme: {
        extend: {
          colors: {
            ...addCommonColors ? commonColors : {},
            ...resolved.colors
          },
          height: {
            divider: `var(--${prefix}-divider-weight)`
          },
          width: {
            divider: `var(--${prefix}-divider-weight)`
          },
          spacing: {
            unit: `var(--${prefix}-spacing-unit)`,
            ...createSpacingUnits(prefix)
          },
          minWidth: {
            ...minSizes
          },
          minHeight: {
            ...minSizes
          },
          fontSize: {
            tiny: [`var(--${prefix}-font-size-tiny)`, `var(--${prefix}-line-height-tiny)`],
            small: [`var(--${prefix}-font-size-small)`, `var(--${prefix}-line-height-small)`],
            medium: [`var(--${prefix}-font-size-medium)`, `var(--${prefix}-line-height-medium)`],
            large: [`var(--${prefix}-font-size-large)`, `var(--${prefix}-line-height-large)`]
          },
          borderRadius: {
            small: `var(--${prefix}-radius-small)`,
            medium: `var(--${prefix}-radius-medium)`,
            large: `var(--${prefix}-radius-large)`
          },
          opacity: {
            disabled: `var(--${prefix}-disabled-opacity)`
          },
          borderWidth: {
            small: `var(--${prefix}-border-width-small)`,
            medium: `var(--${prefix}-border-width-medium)`,
            large: `var(--${prefix}-border-width-large)`,
            1: "1px",
            1.5: "1.5px",
            3: "3px",
            5: "5px"
          },
          boxShadow: {
            small: `var(--${prefix}-box-shadow-small)`,
            medium: `var(--${prefix}-box-shadow-medium)`,
            large: `var(--${prefix}-box-shadow-large)`
          },
          backgroundImage: {
            "stripe-gradient": "linear-gradient(45deg, rgba(0, 0, 0, 0.1) 25%, transparent 25%, transparent 50%, rgba(0, 0, 0, 0.1) 50%, rgba(0, 0, 0, 0.1) 75%, transparent 75%, transparent)"
          },
          transitionDuration: {
            0: "0ms",
            250: "250ms",
            400: "400ms"
          },
          transitionTimingFunction: {
            "soft-spring": "cubic-bezier(0.155, 1.105, 0.295, 1.12)"
          },
          ...animations
        }
      }
    }
  );
};
var nextui = (config = {}) => {
  const {
    themes: themeObject = {},
    defaultTheme = "light",
    layout: userLayout,
    defaultExtendTheme = "light",
    prefix: defaultPrefix = DEFAULT_PREFIX,
    addCommonColors = false
  } = config;
  const userLightColors = get(themeObject, "light.colors", {});
  const userDarkColors = get(themeObject, "dark.colors", {});
  const defaultLayoutObj = userLayout && typeof userLayout === "object" ? deepMerge(defaultLayout, userLayout) : defaultLayout;
  const baseLayouts = {
    light: {
      ...defaultLayoutObj,
      ...lightLayout
    },
    dark: {
      ...defaultLayoutObj,
      ...darkLayout
    }
  };
  let otherThemes = omit(themeObject, ["light", "dark"]) || {};
  forEach(otherThemes, ({ extend, colors, layout }, themeName) => {
    const baseTheme = extend && isBaseTheme(extend) ? extend : defaultExtendTheme;
    if (colors && typeof colors === "object") {
      otherThemes[themeName].colors = deepMerge(semanticColors[baseTheme], colors);
    }
    if (layout && typeof layout === "object") {
      otherThemes[themeName].layout = deepMerge(
        extend ? baseLayouts[extend] : defaultLayoutObj,
        layout
      );
    }
  });
  const light = {
    layout: deepMerge(baseLayouts.light, get(themeObject, "light.layout", {})),
    colors: deepMerge(semanticColors.light, userLightColors)
  };
  const dark = {
    layout: deepMerge(baseLayouts.dark, get(themeObject, "dark.layout", {})),
    colors: deepMerge(semanticColors.dark, userDarkColors)
  };
  const themes = {
    light,
    dark,
    ...otherThemes
  };
  return corePlugin(themes, defaultTheme, defaultPrefix, addCommonColors);
};

export {
  nextui
};
